File: ischia/src/main/java/cz/fb/manaus/ischia/proposer/TradedVolumeProposer.java
package cz.fb.manaus.ischia.proposer;

import cz.fb.manaus.ischia.BackLoserBet;
import cz.fb.manaus.ischia.LayLoserBet;
import cz.fb.manaus.reactor.betting.BetContext;
import cz.fb.manaus.reactor.betting.proposer.PriceProposer;
import cz.fb.manaus.reactor.betting.validator.ValidationResult;
import cz.fb.manaus.reactor.price.PriceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

import java.util.OptionalDouble;


@Component
@BackLoserBet
@LayLoserBet
@Profile("betfair")
public class TradedVolumeProposer implements PriceProposer {

    public static final double REDUCTION_RATE = 0.01;
    @Autowired
    private PriceService priceService;

    @Override
    public ValidationResult validate(BetContext context) {
        if (context.getActualTradedVolume().get().getVolume().size() == 0 || context.getActualTradedVolume().get().getWeightedMean().getAsDouble() > 100) {
            return ValidationResult.REJECT;
        } else {
            return PriceProposer.super.validate(context);
        }
    }

    @Override
    public OptionalDouble getProposedPrice(BetContext context) {
        double weightedMean = context.getActualTradedVolume().get().getWeightedMean().getAsDouble();
        return OptionalDouble.of(priceService.downgrade(weightedMean, REDUCTION_RATE, context.getSide()));
    }


}

File: ischia/src/main/java/cz/fb/manaus/ischia/proposer/TradedVolumeProposer.kt
package cz.fb.manaus.ischia.proposer

import cz.fb.manaus.core.provider.ProviderTag.TradedVolume
import cz.fb.manaus.ischia.BackUniverse
import cz.fb.manaus.ischia.LayUniverse
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.proposer.PriceProposer
import cz.fb.manaus.reactor.betting.validator.ValidationResult
import cz.fb.manaus.reactor.price.Pricing
import org.springframework.stereotype.Component


@Component
@BackUniverse
@LayUniverse
object TradedVolumeProposer : PriceProposer {

    override val tags get() = setOf(TradedVolume)

    override fun validate(event: BetEvent): ValidationResult {
        val tradedVolume = event.metrics.actualTradedVolume ?: error("no vol")
        return if (tradedVolume.volume.isEmpty() || tradedVolume.weightedMean!! > 100) {
            ValidationResult.DROP
        } else {
            super.validate(event)
        }
    }

    override fun getProposedPrice(event: BetEvent): Double {
        val weightedMean = event.metrics.actualTradedVolume!!.weightedMean!!
        return Pricing.downgrade(
                weightedMean, 0.01,
                event.side
        )
    }

}
